eth/downloader/downloader.go

```diff
@@ -1178,23 +1182,75 @@ func (d *Downloader) importBlockResults(results []*fetchResult) error {
 	for i, result := range results {
 		blocks[i] = types.NewBlockWithHeader(result.Header).WithBody(result.Transactions, result.Uncles, result.ExtTransactions, result.SubManifest)
 	}
-	if index, err := d.core.InsertChain(blocks); err != nil {
-		if err.Error() == core.ErrAddedFutureCache.Error() {
-			d.pauseCh<-true // Pause the downloader
-			return nil
-		}
-		if index < len(results) {
-			log.Debug("Downloaded item processing failed", "number", results[index].Header.Number(), "hash", results[index].Header.Hash(), "err", err)
-		} else {
-			// The InsertChain method in core.go will sometimes return an out-of-bounds index,
-			// when it needs to preprocess blocks to import a sidechain.
-			// The importer will put together a new list of blocks to import, which is a superset
-			// of the blocks delivered from the downloader, and the indexing will be off.
-			log.Debug("Downloaded item processing failed on sidechain import", "index", index, "err", err)
+	// Loop until all blocks have been inserted, or the process has been cancelled
+	count := 0
+	waitAppend := common.Hash{}
+	waitSubAppend := common.Hash{}
+	for {
+		select {
+		case <-d.quitCh:
+			return errCancelContentProcessing
+		case <-d.cancelCh:
+			return errCancelContentProcessing
+		default:
+			// Pause if we are waiting on a sub append
+			for !bytes.Equal(waitAppend.Bytes(), common.Hash{}.Bytes()) ||
+				!bytes.Equal(waitSubAppend.Bytes(), common.Hash{}.Bytes()) {
+				select {
+				case <-d.quitCh:
+					return errCancelContentProcessing
+				case hash := <-appendCh:
+					if bytes.Equal(waitAppend.Bytes(), hash.Bytes()) {
+						// Clear the wait states
+						waitAppend = common.Hash{}
+						waitSubAppend = common.Hash{}
+					}
+				case subHash := <-subAppendCh:
+					if bytes.Equal(waitSubAppend.Bytes(), subHash.Bytes()) {
+						// Clear the wait states
+						waitAppend = common.Hash{}
+						waitSubAppend = common.Hash{}
+					}
+				case <-insertWaitTimer.C:
+					waitAppend = common.Hash{}
+					waitSubAppend = common.Hash{} // Clear the wait state
+				}
+			}
+			index, err := d.core.InsertChain(blocks[count:], false)
+			count += index
+			if err != nil {
+				if err.Error() == core.ErrPendingBlock.Error() {
+					// We are waiting on a pending block. This most likely means we are waiting
+					// on a sub or dom append. Wait for either:
+					// 1) our sub to append the give sub parent
+					// 2) our chain to append our parent
+					// 3) the retry timer tick
+					if nodeCtx < common.ZONE_CTX {
+						// Indicate which sub block we are waiting on (possibly)
+						waitSubAppend = blocks[index].ParentHash(nodeCtx + 1)
+					} else {
+						// Otherwise just set it to an impossible value and wait for the timer to try again
+						waitSubAppend = blocks[index].ParentHash()
```

Comment:
shouldn't this be wait on append? not sub append?

Reply:
Yes, good catch. Fixed.